### ISSUE TITLE
Allow CSS to target layout of a certain page

### ISSUE THREAD

**Issue by armano2** [facebook/docusaurus#7592]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Description

I went trough documentation and i'm unable to find an option to disable stickiness of navbar for specific page/s

this refers to class `.navbar--fixed-top`

as an example of page that i'd like to disable stickiness is a typescript-eslint [playground](https://typescript-eslint.io/play/#ts=4.7.1-rc&sourceType=module&code=JYMwBAFALgTgrgUzAMmWAjGA1BglGAbzAF8BYAKCA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugIwHs+SAIZN6AD2LQUySnyaoMAbXDYciaND7RIAGhWqsq7JCHx4fAO4BlTswDm6MADNTyRHqPHT5iwDlYALY86o4u8G4enlDelv5mQjxIAPI8AFaI3KGuiPrYAL65ALqROAQk5FQ09MiMLKxMfLRO5kL49rSSfAGUbgpQ6praJVBlpMgU1HQMzGwNtN3IsG4AJh2aCyiOkANaurl4RGMTVdN1c7BMS4irAG5C0H3bGrvDB+XjlVM1M-WNPEJuWj4RocKhMBwYJ6DSAqAp5IA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

![image](https://user-images.githubusercontent.com/625469/172965234-8b3bc7fc-2737-4a5b-bc6d-53411bdb380d.png)

![image](https://user-images.githubusercontent.com/625469/172965337-7a284749-f42e-46ca-978a-ce9977b3d752.png)

-----------

i'm not sure if this is a documentation request or feature request

### Self-service

- [ ] I'd be willing to address this documentation request myself.

**Comment by slorber** [comment]
Couldn't this work?

https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/website/src/pages/play.tsx

```tsx
function Play(): JSX.Element {
  return (
    <HtmlClassNameProvider className="my-custom-playground-className">
      <Layout title="Playground" description="Playground" noFooter={true}>
        <BrowserOnly fallback={<Loader />}>
          {(): JSX.Element => {
            const Playground = lazy(() => import("../components/Playground"));
            return (
              <Suspense fallback={<Loader />}>
                <Playground />
              </Suspense>
            );
          }}
        </BrowserOnly>
      </Layout>
    </HtmlClassNameProvider>
  );
}
```

This API is not officially documented for now, hopefully we may remove it in the future.

Its only role is to allow composition on avoid Helmet `<Head>` component for `<html className="xyz">`:

```jsx
export function HtmlClassNameProvider({
  className: classNameProp,
  children
}: {
  className: string,
  children: ReactNode
}): JSX.Element {
  const classNameContext = React.useContext(HtmlClassNameContext);
  const className = clsx(classNameContext, classNameProp);
  return (
    <HtmlClassNameContext.Provider value={className}>
      <Head>
        <html className={className} />
      </Head>
      {children}
    </HtmlClassNameContext.Provider>
  );
}
```

If you don't need composition on the playground page, you could as well just use:

```jsx
<Head>
  <html className={className} />
</Head>
```

**Comment by armano2** [comment]
@slorber yes your solution is working, but at the end i decided to do it differently, by changing how tooltips are rendered in editor

**Comment by slorber** [comment]
👍 

So I'll close this issue for now

**Comment by Josh-Cena** [comment]
This is a good solution for React pages, but not for Markdown pages where you don't have access to the wrapping component. In the long term I think we should have a route-specific class name for each page.

**Comment by slorber** [comment]
> This is a good solution for React pages, but not for Markdown pages where you don't have access to the wrapping component. In the long term I think we should have a route-specific class name for each page.

We have doc id/version for each md doc.

And it looks like frontmatter `wrapperClassName` is wired to HTML className already for md pages.

I think only blog posts do not easily allow to be targeted with granularity.

**Comment by Josh-Cena** [comment]
Ah yes, forgot about that. Correct, only blog posts can't be differentiated

**Comment by slorber** [comment]
I think we can re-open and do this:

- md page `frontMatter.wrapperClassName` => apply to Layout wrapperClassName (instead of html, for consistency with DocPage)
- md page `frontMatter.htmlClassName` (new)
- blog => find a way to generate a unique html className per post

There are other cases (like category index...), but I doubt they would be as useful as docs/blog/pages

**Comment by Michaelsulistio** [comment]
@slorber Hi Slorber, I'm running into this exact issue, where I want a way to target styling to a specific doc file.

I see above, you talk about md.page `frontMatter.wrapperClassName` or `htmlClassName`. If I understand correctly, with these, I'd be able to globally target the provided `wrapperClassName` and `htmlClassName` and custom style a Markdown page.

Have these been added yet?

**Comment by slorber** [comment]
@Michaelsulistio yes, give it a try and see what elements in the DOM permits you to target that specific page.

> where I want a way to target styling to a specific doc file.

If it's a doc page (ie of the docs plugin) we already add many metadata at the top to help you target it individually

For example, this page: https://docusaurus.io/docs/configuration has the following html classname:

```html
<html lang="en" dir="ltr" class="docs-wrapper plugin-docs plugin-id-default docs-version-2.4.0 docs-doc-page docs-doc-id-configuration">
```

This should be largely enough to target it.

If you want more help, please provide a concrete example in a repro with the page you want to target and I can tell you how to do so.

---

This issue remain open because there are still some things that you can't target easily (like blog posts).  See https://github.com/facebook/docusaurus/issues/7592#issuecomment-1157553912)

But most of it has been implemented already, and you should be able to achieve what you want here.

**Comment by nektro** [comment]
does the `@docusaurus/Head` method work if the goal is to include a separate processed `/src/css/custom2.css` only on a specific `/src/pages/foo.mdx` page?

**Comment by slorber** [comment]
> does the @docusaurus/Head method work if the goal is to include a separate processed /src/css/custom2.css only on a specific /src/pages/foo.mdx page?

`<Head>` works the way it's documented (https://github.com/staylor/react-helmet-async) anywhere JSX works, including MDX (which is compiled to JSX in the end). "works" is ambiguous so if you read the doc and it doesn't work as documented, you should open an issue to the package repo directly.
